pensieve/config.py

```diff
@@ -0,0 +1,309 @@
+    name: str
+
+    def resolve(self, spec: "AnalysisSpec") -> mozanalysis.metrics.DataSource:
+        search = mozanalysis.metrics.desktop
+        return _lookup_name(
+            name=self.name,
+            klass=mozanalysis.metrics.DataSource,
+            spec=spec,
+            module=search,
+            definitions=spec.data_sources.definitions,
+        )
+
+
+_converter.register_structure_hook(
+    DataSourceReference, lambda obj, _type: DataSourceReference(name=obj)
+)
+
+
+@attr.s(auto_attribs=True)
+class ExperimentConfiguration:
+    """Represents the configuration of an experiment for analysis."""
+
+    experiment_spec: "ExperimentSpec"
+    experimenter_experiment: pensieve.experimenter.Experiment
+
+    def __getattr__(self, name):
+        if hasattr(self.experiment_spec, name):
+            return getattr(self.experiment_spec, name)
+        return getattr(self.experimenter_experiment, name)
+
+
+@attr.s(auto_attribs=True)
+class ExperimentSpec:
+    """Describes the interface for overriding experiment details."""
+
+    # TODO: Expand this list.
+    enrollment_query: Optional[str] = None
+
+    def resolve(
+        self, spec: "AnalysisSpec", experimenter: pensieve.experimenter.Experiment
+    ) -> ExperimentConfiguration:
+        return ExperimentConfiguration(self, experimenter)
+
+
+@attr.s(auto_attribs=True)
+class MetricDefinition:
+    """Describes the interface for defining a metric in configuration.
+
+    The `select_expression` of the metric may use Jinja2 template syntax to refer to the
+    aggregation helper functions defined in `mozanalysis.metrics`, like
+        '{{agg_any("payload.processes.scalars.some_boolean_thing")}}'
+    """
+
+    name: str  # implicit in configuration
+    select_expression: str
+    data_source: DataSourceReference
+
+    def resolve(self, spec: "AnalysisSpec") -> mozanalysis.metrics.Metric:
+        select_expression = _metrics_environment.from_string(self.select_expression).render()
+
+        return mozanalysis.metrics.Metric(
+            name=self.name,
+            data_source=self.data_source.resolve(spec),
+            select_expr=select_expression,
+        )
+
+
+@attr.s(auto_attribs=True)
+class MetricsConfiguration:
+    """Contains the metrics configured for each analysis window of an experiment."""
+
+    daily: Set[mozanalysis.metrics.Metric]
+    weekly: Set[mozanalysis.metrics.Metric]
+    overall: Set[mozanalysis.metrics.Metric]
+
+
+@attr.s(auto_attribs=True)
+class MetricsSpec:
+    """Describes the interface for the metrics section in configuration."""
+
+    daily: List[MetricReference] = attr.Factory(list)
+    weekly: List[MetricReference] = attr.Factory(list)
+    overall: List[MetricReference] = attr.Factory(list)
+
+    definitions: Dict[str, MetricDefinition] = attr.Factory(dict)
+
+    @classmethod
+    def from_dict(cls, d: dict) -> "MetricsSpec":
+        params: Dict[str, Any] = {}
+        known_keys = ("daily", "weekly", "overall")
```

Comment:
I'm wondering if it makes sense to make use of the `AnalysisPeriod` enum also in the config. 
This could for example be rewritten to `for period in AnalysisPeriod:`

Reply:
Yes, good call. It seems like we want to treat the keys of `MetricsConfiguration` as a dict so often I kind of wonder if it shouldn't just be a `Dict[AnalysisPeriod, Metric]` but maybe that's a question for another day.